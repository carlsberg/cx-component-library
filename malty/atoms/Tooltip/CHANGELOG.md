# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- Failling tooltip tests ([#210](https://github.com/CarlsbergGBS/cx-component-library/issues/210)) ([d040100](https://github.com/CarlsbergGBS/cx-component-library/commit/d0401006ed0895dfdd945cc6cd5ec1c269d2c35f))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- Tooltip hover and positioning ([#208](https://github.com/CarlsbergGBS/cx-component-library/issues/208)) ([d151215](https://github.com/CarlsbergGBS/cx-component-library/commit/d15121511693eeeb9d13385f27124deec3fa8f8c))
- Tooltip positioning and onClose event ([#198](https://github.com/CarlsbergGBS/cx-component-library/issues/198)) ([2b3788b](https://github.com/CarlsbergGBS/cx-component-library/commit/2b3788bd71d6870a8e3663b2250bd2b87d737791))
- **tooltip:** Force recalculation of tooltip position after opening it ([#492](https://github.com/CarlsbergGBS/cx-component-library/issues/492)) ([3060b94](https://github.com/CarlsbergGBS/cx-component-library/commit/3060b94c4b335a92c52490f94b8a475b4e0b073d))
- Update components to have a function declaration and remove unneeded fragments ([#502](https://github.com/CarlsbergGBS/cx-component-library/issues/502)) ([b470095](https://github.com/CarlsbergGBS/cx-component-library/commit/b4700954f4e1ae9c7e547f4de37ac2b15a9d55ea))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **chore:** [DSM-714] add lint and format commands ([#705](https://github.com/CarlsbergGBS/cx-component-library/issues/705)) ([5b8c3bf](https://github.com/CarlsbergGBS/cx-component-library/commit/5b8c3bf45dc5c32cd52724a172578b36d93427a5))
- **DSM-285:** table ([#442](https://github.com/CarlsbergGBS/cx-component-library/issues/442)) ([8f81f8d](https://github.com/CarlsbergGBS/cx-component-library/commit/8f81f8d3436af86fd023d9737245a76a142ad51f))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** Remove `react/jsx-props-no-spreading` comments ([#558](https://github.com/CarlsbergGBS/cx-component-library/issues/558)) ([823444c](https://github.com/CarlsbergGBS/cx-component-library/commit/823444ce0800bc7f49b584de59d247eb8497f559))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
