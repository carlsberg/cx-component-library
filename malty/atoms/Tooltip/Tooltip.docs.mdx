---
labels: ['react', 'typescript', 'text']
description: 'A Tooltip component.'
---

import { Tooltip, Position } from '.';

Text example:

A simple top Tooltip component with some text

```js live
<Tooltip position={Position.Top} isOpen>
  This is the content
</Tooltip>
```
