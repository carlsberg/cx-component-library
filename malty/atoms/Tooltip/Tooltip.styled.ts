import styled, { keyframes } from 'styled-components';

// Event
const fadeInAnimation = keyframes`
  0% {
    opacity: 0.3
  }
  100% {
    opacity: 1
  }
`;

export const StyledArrow = styled.div<{
  isDark: boolean;
}>`
  border-style: solid;
  border-color: transparent;
`;

export const StyledTooltipWrapper = styled.div<{
  isDark: boolean;
}>`
  font-family: inherit;
  background-color: ${({ isDark, theme }) =>
    isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
  min-width: ${({ theme }) => theme.sizes['4xl'].value};
  box-shadow: 0 0 5px rgba(0, 0, 0, 0.15);
  animation-name: ${fadeInAnimation};
  animation-duration: 0.5s;

  &[data-popper-placement^='top'] ${StyledArrow} {
    bottom: -${({ theme }) => theme.sizes['3xs'].value};
    border-width: ${({ theme }) => theme.sizes['3xs'].value} ${({ theme }) => theme.sizes['3xs'].value} 0
      ${({ theme }) => theme.sizes['3xs'].value};
    border-top-color: ${({ theme, isDark }) =>
      isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
  }

  &[data-popper-placement^='bottom'] ${StyledArrow} {
    top: -${({ theme }) => theme.sizes['3xs'].value};
    border-width: 0 ${({ theme }) => theme.sizes['3xs'].value} ${({ theme }) => theme.sizes['3xs'].value}
      ${({ theme }) => theme.sizes['3xs'].value};
    border-bottom-color: ${({ theme, isDark }) =>
      isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
  }

  &[data-popper-placement^='left'] ${StyledArrow} {
    right: -${({ theme }) => theme.sizes['3xs'].value};
    border-width: ${({ theme }) => theme.sizes['3xs'].value} 0 ${({ theme }) => theme.sizes['3xs'].value}
      ${({ theme }) => theme.sizes['3xs'].value};
    border-left-color: ${({ theme, isDark }) =>
      isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
  }

  &[data-popper-placement^='right'] ${StyledArrow} {
    left: -${({ theme }) => theme.sizes['3xs'].value};
    border-width: ${({ theme }) => theme.sizes['3xs'].value} ${({ theme }) => theme.sizes['3xs'].value}
      ${({ theme }) => theme.sizes['3xs'].value} 0;
    border-right-color: ${({ theme, isDark }) =>
      isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
  }
`;

export const StyledTooltip = styled.div`
  padding: ${({ theme }) => `${theme.sizes['4xs'].value} ${theme.sizes['2xs'].value}`};
`;

// export const StyledTooltipPositionTopCenter = styled(StyledTooltip)`
//   bottom: calc(100% + ${({ theme }) => theme.sizes['2xs'].value});
//   left: 50%;
//   transform: translate(-50%, 0);
//   margin: ${({ anchorOffset }) => `0 0 ${anchorOffset.vertical}px -${anchorOffset.horizontal / 2}px`};

//   &::before {
//     left: 50%;
//     right: auto;
//     transform: translateX(-50%);
//     bottom: -${({ theme }) => theme.sizes['3xs'].value};
//     border-width: ${({ theme }) => theme.sizes['3xs'].value} ${({ theme }) => theme.sizes['3xs'].value} 0
//       ${({ theme }) => theme.sizes['3xs'].value};
//     border-top-color: ${({ theme, isDark }) =>
//       isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
//   }
// `;

// export const StyledTooltipPositionTopLeft = styled(StyledTooltipPositionTopCenter)`
//   margin: ${({ anchorOffset }) => `0 0 ${anchorOffset.vertical}px -${anchorOffset.horizontal / 1.5}px`};
//   &::before {
//     left: ${({ theme }) => theme.sizes['2xs'].value};
//     right: auto;
//     transform: none;
//   }
// `;

// export const StyledTooltipPositionTopRight = styled(StyledTooltipPositionTopCenter)`
//   margin: ${({ anchorOffset }) => `0 0 ${anchorOffset.vertical}px -${anchorOffset.horizontal / 3}px`};
//   &::before {
//     left: auto;
//     right: ${({ theme }) => theme.sizes['2xs'].value};
//   }
// `;

// export const StyledTooltipPositionBottomCenter = styled(StyledTooltip)`
//   top: calc(100% + ${({ theme }) => theme.sizes.s.value});
//   left: 50%;
//   transform: translate(-50%, 0);
//   margin: ${({ anchorOffset }) => `0 0 0 -${anchorOffset.horizontal / 2}px`};

//   &::before {
//     top: -${({ theme }) => theme.sizes['3xs'].value};
//     left: 50%;
//     right: auto;
//     transform: translateX(-50%);
//     border-width: 0 ${({ theme }) => theme.sizes['3xs'].value} ${({ theme }) => theme.sizes['3xs'].value}
//       ${({ theme }) => theme.sizes['3xs'].value};
//     border-bottom-color: ${({ theme, isDark }) =>
//       isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
//   }
// `;

// export const StyledTooltipPositionBottomLeft = styled(StyledTooltipPositionBottomCenter)`
//   margin: ${({ anchorOffset }) => `0 0 0 -${anchorOffset.horizontal / 1.5}px`};
//   &::before {
//     left: ${({ theme }) => theme.sizes['2xs'].value};
//     right: auto;
//     transform: none;
//   }
// `;

// export const StyledTooltipPositionBottomRight = styled(StyledTooltipPositionBottomCenter)`
//   margin: ${({ anchorOffset }) => `0 0 0 -${anchorOffset.horizontal / 3}px`};
//   &::before {
//     left: auto;
//     right: ${({ theme }) => theme.sizes['2xs'].value};
//   }
// `;

// export const StyledTooltipPositionRight = styled(StyledTooltip)`
//   top: 50%;
//   left: calc(100% + ${({ theme }) => theme.sizes.s.value});
//   right: auto;
//   transform: translate(0, -50%);
//   margin: ${({ anchorOffset }) => `-${anchorOffset.vertical / 2}px 0 0 0`};

//   &::before {
//     top: 50%;
//     left: -${({ theme }) => theme.sizes['3xs'].value};
//     transform: translateY(-50%);
//     border-width: ${({ theme }) => theme.sizes['3xs'].value} ${({ theme }) => theme.sizes['3xs'].value}
//       ${({ theme }) => theme.sizes['3xs'].value} 0;
//     border-right-color: ${({ theme, isDark }) =>
//       isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
//   }
// `;

// export const StyledTooltipPositionLeft = styled(StyledTooltip)`
//   top: 50%;
//   left: auto;
//   right: calc(100% + ${({ theme }) => theme.sizes.s.value});
//   transform: translate(0, -50%);
//   margin: ${({ anchorOffset }) => `-${anchorOffset.vertical / 2}px ${anchorOffset.horizontal}px 0 0`};

//   &::before {
//     top: 50%;
//     right: -${({ theme }) => theme.sizes['3xs'].value};
//     transform: translateY(-50%);
//     border-width: ${({ theme }) => theme.sizes['3xs'].value} 0 ${({ theme }) => theme.sizes['3xs'].value}
//       ${({ theme }) => theme.sizes['3xs'].value};
//     border-left-color: ${({ theme, isDark }) =>
//       isDark ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[20].value};
//   }
// `;
