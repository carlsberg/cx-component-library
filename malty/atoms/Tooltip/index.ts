export { Tooltip } from './Tooltip';
export { TooltipPlacement, TooltipPositionStrategy, TooltipToggle } from './Tooltip.types';
export type { TooltipProps } from './Tooltip.types';
