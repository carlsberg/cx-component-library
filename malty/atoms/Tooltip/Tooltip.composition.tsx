import { MaltyThemeProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useState } from 'react';
import { Tooltip } from './Tooltip';
import { Position } from './Tooltip.types';

// sets the Component preview in gallery view
export const BasicTooltip = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <MaltyThemeProvider theme="global">
      <Tooltip position={Position.Bottom} content={<span>Contents</span>} isOpen>
        <button type="button" onClick={() => setIsOpen(!isOpen)}>
          Toggle
        </button>
      </Tooltip>
    </MaltyThemeProvider>
  );
};
