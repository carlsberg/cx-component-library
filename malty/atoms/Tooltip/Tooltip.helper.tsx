import { useCallback, useEffect, useRef, useState } from 'react';
import { usePopper } from 'react-popper';
import { TooltipPositionStrategy, TooltipToggle, UseTooltipProps } from './Tooltip.types';

export const useToolTip = ({
  positionStrategy = TooltipPositionStrategy.Absolute,
  placement,
  toggleType,
  isOpenProp,
  onClose,
  autoHideDuration,
  tooltipId
}: UseTooltipProps) => {
  const didMountRef = useRef(false);
  const [isOpen, setIsOpen] = useState(!!isOpenProp);
  const [referenceElement, setReferenceElement] = useState<HTMLElement | null>(null);
  const [popperElement, setPopperElement] = useState<HTMLElement | null>(null);
  const [arrowElement, setArrowElement] = useState<HTMLElement | null>(null);
  const { styles, attributes, update } = usePopper(referenceElement, popperElement, {
    strategy: positionStrategy,
    placement,
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [-8, 8]
        }
      },
      {
        name: 'arrow',
        options: {
          element: arrowElement
        }
      }
    ]
  });

  const autoHideTimer = useRef<number | NodeJS.Timeout | null>(null);

  const clearAutoHideTimer = useCallback(() => {
    if (typeof autoHideTimer?.current === 'number') {
      clearTimeout(autoHideTimer.current);
    }
  }, []);

  const setTooltipOpen = useCallback(
    (open: boolean) => {
      if (typeof isOpenProp !== 'boolean') {
        clearAutoHideTimer();
        setIsOpen(open);
      }
    },
    [isOpenProp, clearAutoHideTimer]
  );

  const startAutoHideTimer = useCallback(() => {
    if (toggleType === TooltipToggle.Event && typeof isOpenProp !== 'boolean') {
      setTooltipOpen(true);
      autoHideTimer.current = setTimeout(() => {
        setIsOpen(false);
      }, autoHideDuration);
    }
  }, [toggleType, isOpenProp, autoHideDuration, setTooltipOpen]);

  useEffect(() => {
    const validateEventSource = (ev: Event) => (ev as CustomEvent).detail === tooltipId;
    const onStartAutoHideTimerEvent = (ev: Event) => {
      if (validateEventSource(ev)) {
        startAutoHideTimer();
      }
    };

    const onOpenTooltipEvent = (ev: Event) => {
      if (validateEventSource(ev)) {
        setTooltipOpen(true);
      }
    };

    const onCloseTooltipEvent = (ev: Event) => {
      if (validateEventSource(ev)) {
        setTooltipOpen(false);
      }
    };
    window.addEventListener(START_TOOLTIP_TIMER_EVENT, onStartAutoHideTimerEvent);
    window.addEventListener(OPEN_TOOLTIP_EVENT, onOpenTooltipEvent);
    window.addEventListener(CLOSE_TOOLTIP_EVENT, onCloseTooltipEvent);
    return () => {
      clearAutoHideTimer();
      window.removeEventListener(START_TOOLTIP_TIMER_EVENT, onStartAutoHideTimerEvent);
      window.removeEventListener(OPEN_TOOLTIP_EVENT, onOpenTooltipEvent);
      window.removeEventListener(CLOSE_TOOLTIP_EVENT, onCloseTooltipEvent);
    };
  }, [startAutoHideTimer, setTooltipOpen, clearAutoHideTimer]);

  useEffect(() => {
    if (toggleType === TooltipToggle.Hover) {
      const handleAnchorMouseEnter = () => {
        setTooltipOpen(true);
      };

      const handleAnchorMouseLeave = () => {
        setTooltipOpen(false);
      };

      referenceElement?.addEventListener('mouseenter', handleAnchorMouseEnter);
      referenceElement?.addEventListener('mouseleave', handleAnchorMouseLeave);

      return () => {
        referenceElement?.removeEventListener('mouseenter', handleAnchorMouseEnter);
        referenceElement?.removeEventListener('mouseleave', handleAnchorMouseLeave);
      };
    }

    return () => null;
  }, [referenceElement, toggleType, setTooltipOpen]);

  useEffect(() => {
    if (toggleType === TooltipToggle.Click) {
      const handleTooltipToggle = () => {
        setTooltipOpen(!isOpen);
      };

      referenceElement?.addEventListener('click', handleTooltipToggle);

      return () => {
        referenceElement?.removeEventListener('click', handleTooltipToggle);
      };
    }

    return () => null;
  }, [referenceElement, toggleType, isOpen, setTooltipOpen]);

  useEffect(() => {
    if (typeof isOpenProp === 'boolean' && isOpenProp !== isOpen) {
      setIsOpen(isOpenProp);
    }
  }, [isOpenProp]);

  useEffect(() => {
    if (didMountRef.current && !isOpen) {
      onClose?.();
    }
    didMountRef.current = true;
  }, [isOpen, onClose]);

  return {
    styles,
    attributes,
    isOpen,
    setTooltipOpen,
    setReferenceElement,
    setPopperElement,
    setArrowElement,
    startAutoHideTimer,
    updateTooltipPosition: update
  };
};

export const START_TOOLTIP_TIMER_EVENT = 'START_TOOLTIP_TIMER_EVENT';
export const OPEN_TOOLTIP_EVENT = 'OPEN_TOOLTIP_EVENT';
export const CLOSE_TOOLTIP_EVENT = 'CLOSE_TOOLTIP_EVENT';
