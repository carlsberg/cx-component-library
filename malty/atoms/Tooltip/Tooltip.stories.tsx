import { Image } from '@carlsberggroup/malty.atoms.image';
import { Text, TextColor, TextStyle } from '@carlsberggroup/malty.atoms.text';
import { Meta, Story } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import { Tooltip as TooltipComponent } from './Tooltip';
import {
  TooltipPlacement,
  TooltipPositionStrategy,
  TooltipProps,
  TooltipToggle,
} from './Tooltip.types';

const StyledContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 200px;

  span {
    width: 0;
  }

  p {
    margin-block-start: 0;
    margin-block-end: 0;
  }
`;

export default {
  title: 'Information/Tooltip',
  component: TooltipComponent,
  parameters: {
    importObject: 'Tooltip',
    importPath: '@carlsberggroup/malty.atoms.tooltip',
    variants: ['dark', 'light'],
  },
  argTypes: {
    tooltipId: {
      control: 'text',
      description: 'Tooltip Id',
      table: { defaultValue: { summary: 'tooltip' } },
      defaultValue: 'tooltip',
    },
    triggerComponent: {
      control: {
        disable: true,
      },
      description:
        "Anchor element to have Tooltip anchor to. The position is based on this element. If no anchor provided the Tooltip will show in it's corresponding position on the markup, and anchor on itself.",
    },
    isOpen: {
      description: 'Tooltip visibility. Overrides all variants actions that toggle visibility',
      control: 'boolean',
      table: { defaultValue: { summary: 'undefined' } },
    },
    placement: {
      description: 'Tooltip position.',
      options: Object.values(TooltipPlacement),
      table: { defaultValue: { summary: 'TooltipPlacement.Bottom' } },
      control: {
        type: 'select',
        label: Object.keys(TooltipPlacement),
      },
      defaultValue: 'bottom',
    },

    positionStrategy: {
      description: 'Tooltip positioning strategy.',
      options: Object.values(TooltipPositionStrategy),
      mapping: TooltipPositionStrategy,
      table: { defaultValue: { summary: 'TooltipPositionStrategy.ABSOLUTE' } },
      control: {
        type: 'select',
        label: Object.keys(TooltipPositionStrategy),
      },
      defaultValue: 'absolute',
    },
    children: {
      description:
        'Content for the Tooltip, can be a `string`, a `React Element` or just simply `HTML`.',
      table: {
        type: {
          summary: 'JSX.Element',
        },
      },
      control: 'text',
    },
    toggle: {
      description: 'Expected Tooltip behaviour for trigger.',
      options: Object.values(TooltipToggle),
      mapping: TooltipToggle,
      table: { defaultValue: { summary: 'TooltipToggle.Hover' } },
      control: {
        type: 'select',
        label: Object.keys(TooltipToggle),
      },
      defaultValue: 'Hover',
    },
    isDark: {
      description: 'Dark theme for the Tooltip.',
      control: 'boolean',
      table: { defaultValue: { summary: 'true' } },
    },
    dataTestId: {
      control: 'text',
      description: 'Tooltip data-testid',
      table: { defaultValue: { summary: 'none' } },
    },
    autoHideDuration: {
      control: 'number',
      description: 'Set auto hide duration - available only for `Event` toggle',
      table: { defaultValue: { summary: '5000' } },
    },
    onClose: {
      description: 'Function to be executed when tooltip state is changed to hidden.',
    },
  },
} as Meta;

const Template: Story<TooltipProps> = ({
  placement,
  toggle,
  isDark,
  dataTestId,
  autoHideDuration,
  isOpen,
  children,
}: TooltipProps) => {
  const tooltipTextColor = isDark ? TextColor.White : TextColor.DigitalBlack;
  const renderTooltipEventToggle = () => (
    <Text textStyle={TextStyle.TinyBold} color={tooltipTextColor}>
      {children}
    </Text>
  );
  const toggleVisibility = (open: boolean) => () => {
    if (open) {
      TooltipComponent.openTooltip('tooltip');
    } else {
      TooltipComponent.closeTooltip('tooltip');
    }
  };

  const startTimer = () => {
    TooltipComponent.startTooltipTimer('tooltip');
  };

  return (
    <StyledContainer>
      <TooltipComponent
        tooltipId="tooltip"
        triggerComponent={(setTriggerElement) => (
          <p ref={setTriggerElement}>Choose your toggle control and play with me!!!</p>
        )}
        placement={placement}
        toggle={toggle}
        isOpen={isOpen}
        isDark={isDark}
        autoHideDuration={autoHideDuration}
        dataTestId={dataTestId}
      >
        {toggle === TooltipToggle.Event ? renderTooltipEventToggle() : children}
      </TooltipComponent>

      {toggle === TooltipToggle.Event && (
        <p>
          <button onClick={toggleVisibility(true)} type="button">
            Show
          </button>
          <button onClick={toggleVisibility(false)} type="button">
            Hide
          </button>
          <button onClick={startTimer} type="button">
            Open with timer
          </button>
        </p>
      )}
    </StyledContainer>
  );
};

export const Tooltip = Template.bind({});

const params = new URLSearchParams(window.location.search);
const variant = params.get('variant');

switch (variant) {
  case 'dark':
    Tooltip.args = {
      placement: TooltipPlacement.Top,
      toggle: TooltipToggle.Hover,
      dataTestId: 'tooltip',
      children: (
        <div style={{ padding: '5px 0 0 0' }}>
          <Image src="https://via.placeholder.com/90x?text=Any+HTML" />
        </div>
      ),
      isDark: true,
      tooltipId: 'tooltip',
      triggerComponent: (setTriggerElement) => (
        <p ref={setTriggerElement}>Choose your toggle control and play with me!!!</p>
      ),
    };
    break;

  default:
    Tooltip.args = {
      placement: TooltipPlacement.Top,
      toggle: TooltipToggle.Hover,
      dataTestId: 'tooltip',
      children: 'A simple Tooltip content with some text. Thanks for open me!',
      isDark: false,
      tooltipId: 'tooltip',
      triggerComponent: (setTriggerElement) => (
        <p ref={setTriggerElement}>Choose your toggle control and play with me!!!</p>
      ),
    };
    break;
}
