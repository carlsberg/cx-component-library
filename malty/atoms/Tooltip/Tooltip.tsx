import { Text, TextColor, TextStyle } from '@carlsberggroup/malty.atoms.text';
import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext, useLayoutEffect } from 'react';
import { ThemeContext } from 'styled-components';
import { CLOSE_TOOLTIP_EVENT, OPEN_TOOLTIP_EVENT, START_TOOLTIP_TIMER_EVENT, useToolTip } from './Tooltip.helper';
import { StyledArrow, StyledTooltip, StyledTooltipWrapper } from './Tooltip.styled';
import { TooltipPositionStrategy, TooltipToggle, TooltipType } from './Tooltip.types';

const Tooltip: TooltipType = function Tooltip({
  placement,
  toggle = TooltipToggle.Hover,
  isDark = true,
  dataTestId,
  autoHideDuration = 5000,
  onClose,
  isOpen: isOpenProp,
  triggerComponent,
  tooltipId,
  positionStrategy = TooltipPositionStrategy.Absolute,
  children
}) {
  const theme = useContext(ThemeContext) || defaultTheme;
  const {
    isOpen,
    startAutoHideTimer,
    setTooltipOpen,
    setReferenceElement,
    setPopperElement,
    setArrowElement,
    updateTooltipPosition,
    styles,
    attributes
  } = useToolTip({
    placement,
    isOpenProp,
    autoHideDuration,
    toggleType: toggle,
    onClose,
    tooltipId,
    positionStrategy
  });

  useLayoutEffect(() => {
    if (isOpen) {
      updateTooltipPosition?.();
    }
  }, [isOpen, updateTooltipPosition]);

  const handleTooltipMouseEnter = () => {
    if (toggle === TooltipToggle.Event) {
      setTooltipOpen(true);
    }
  };

  const handleTooltipMouseOut = () => {
    if (toggle === TooltipToggle.Event) {
      startAutoHideTimer();
    }
  };

  const renderChildren = () => {
    if (typeof children !== 'string') {
      return children;
    }

    const tooltipTextColor = isDark ? TextColor.White : TextColor.DigitalBlack;

    return (
      <Text textStyle={TextStyle.TinyBold} color={tooltipTextColor}>
        {children}
      </Text>
    );
  };

  return (
    <>
      {triggerComponent(setReferenceElement)}

      <StyledTooltipWrapper
        {...attributes.popper}
        ref={setPopperElement}
        style={styles.popper}
        data-testid={dataTestId}
        theme={theme}
        isDark={isDark}
        isOpen={isOpen}
        onMouseEnter={handleTooltipMouseEnter}
        onMouseOut={handleTooltipMouseOut}
      >
        <StyledTooltip theme={theme}>{renderChildren()}</StyledTooltip>
        <StyledArrow data-popper-arrow theme={theme} isDark={isDark} ref={setArrowElement} style={styles.arrow} />
      </StyledTooltipWrapper>
    </>
  );
};

Tooltip.startTooltipTimer = (triggerElement) => {
  const event = new window.CustomEvent(START_TOOLTIP_TIMER_EVENT, { detail: triggerElement });
  window.dispatchEvent(event);
};

Tooltip.openTooltip = (triggerElement) => {
  const event = new window.CustomEvent(OPEN_TOOLTIP_EVENT, { detail: triggerElement });
  window.dispatchEvent(event);
};

Tooltip.closeTooltip = (triggerElement) => {
  const event = new window.CustomEvent(CLOSE_TOOLTIP_EVENT, { detail: triggerElement });
  window.dispatchEvent(event);
};
export { Tooltip };
