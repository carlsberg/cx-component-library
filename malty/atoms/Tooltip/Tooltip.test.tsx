import { fireEvent, jsonRenderer, render, screen, waitForElementToBeRemoved } from '@carlsberggroup/malty.utils.test';
import React, { createRef, RefObject } from 'react';
import { act } from 'react-dom/test-utils';
import { Tooltip } from '.';
import { TooltipPosition, TooltipToggle } from './Tooltip.types';

const anchorAction = jest.fn();
const tooltipAnchorRef: RefObject<HTMLAnchorElement> = createRef();

const renderTooltip = (options = { toggleType: TooltipToggle.Click }) =>
  render(
    <div>
      <a onClick={anchorAction} ref={tooltipAnchorRef}>
        Tooltip Anchor
      </a>
      <Tooltip
        anchorRef={tooltipAnchorRef}
        position={TooltipPosition.BottomCenter}
        dataTestId="tooltip-test"
        toggle={options.toggleType}
        autoHideDuration={3000}
      >
        <button type="button">Button inside Tooltip</button>
      </Tooltip>
    </div>
  );

describe('Tooltip', () => {
  it('should matches snapshot', () => {
    const view = jsonRenderer(
      <div>
        <a onClick={anchorAction} ref={tooltipAnchorRef}>
          Tooltip Anchor
        </a>
        <Tooltip
          anchorRef={tooltipAnchorRef}
          position={TooltipPosition.BottomCenter}
          toggle={TooltipToggle.Click}
          autoHideDuration={3000}
        >
          <button type="button">Button inside Tooltip</button>
        </Tooltip>
      </div>
    );
    expect(view).toMatchSnapshot();
  });

  it('should show tooltip once anchor is clicked', () => {
    renderTooltip({ toggleType: TooltipToggle.Click });

    // TooltipToggle.Click - initial state is hidden
    expect(screen.queryByText('Button inside Tooltip')).not.toBeInTheDocument();

    // show tooltip
    fireEvent.click(screen.getByText('Tooltip Anchor'));
    expect(anchorAction).toHaveBeenCalled();
    expect(screen.getByRole('button', { name: 'Button inside Tooltip' })).toBeInTheDocument();

    // hide tooltip
    fireEvent.click(screen.getByText('Tooltip Anchor'));
    expect(anchorAction).toHaveBeenCalled();
    expect(screen.queryByText('Button inside Tooltip')).not.toBeInTheDocument();
  });

  it('should show tooltip once anchor is hovered', () => {
    renderTooltip({ toggleType: TooltipToggle.Hover });

    // TooltipToggle.Hover - initial state is hidden
    expect(screen.queryByText('Button inside Tooltip')).not.toBeInTheDocument();

    // show tooltip
    fireEvent.mouseEnter(screen.getByText('Tooltip Anchor'));
    expect(anchorAction).toHaveBeenCalled();
    expect(screen.getByRole('button', { name: 'Button inside Tooltip' })).toBeInTheDocument();

    // hide tooltip
    fireEvent.mouseLeave(screen.getByText('Tooltip Anchor'));
    expect(anchorAction).toHaveBeenCalled();
    expect(screen.queryByText('Button inside Tooltip')).not.toBeInTheDocument();
  });

  it('should hide tooltip after finish autoHideDuration', async () => {
    renderTooltip({ toggleType: TooltipToggle.Event });

    // TooltipToggle.Event - initial state is hidden
    expect(screen.queryByText('Button inside Tooltip')).not.toBeInTheDocument();

    await act(async () => {
      Tooltip.openTooltip(tooltipAnchorRef);
    });
    // should have opened
    const innerButton = screen.getByRole('button', { name: 'Button inside Tooltip' });
    expect(innerButton).toBeInTheDocument();

    await act(async () => {
      Tooltip.closeTooltip(tooltipAnchorRef);
    });
    // should have closed
    expect(innerButton).not.toBeInTheDocument();

    await act(async () => {
      Tooltip.startTooltipTimer(tooltipAnchorRef);
    });
    // should open again
    const newInnerButton = screen.getByRole('button', { name: 'Button inside Tooltip' });
    expect(newInnerButton).toBeInTheDocument();

    // should hide tooltip after 3 seconds
    await waitForElementToBeRemoved(() => screen.queryByText('Button inside Tooltip'), { timeout: 3000 });
  });
});
