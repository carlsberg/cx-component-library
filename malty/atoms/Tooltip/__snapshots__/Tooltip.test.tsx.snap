// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip matches snapshot 1`] = `
<div
  className="sc-dlnjwi bVqWlg"
  style={
    Object {
      "--color": "dodgerblue",
    }
  }
>
  <div
    className="sc-gtsrHT fvwFPV"
    open={true}
  >
    <p
      className="sc-bdnxRM eWcsub"
    >
      <button
        type="button"
      >
        Test
      </button>
    </p>
  </div>
</div>
`;
